File: probs/26159.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**8)

input = open('input.txt', 'r').readline

from heapq import heappop, heappush, heapify

MOD = int(1e9)+7

def dfs(u, p):
    for v in G[u]:
        if v != p:
            dfs(v, u)
            C[u] += C[v]
            heappush(B, -C[v] * (N - C[v]))

N = int(input())
G = [[] for _ in range(N+1)]
for _ in range(N-1):
    u, v = map(int, input().split())
    G[u].append(v)
    G[v].append(u)
A = list(map(int, input().split()))
heapify(A)

B, C = [], [1]*(N+1)
dfs(1, 0)

ans = 0
while A:
    ans -= heappop(A) * heappop(B)
    ans %= MOD
print(ans)

File: probs/26159.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

const int MAX = 1e5+1,
          MOD = 1e9+7;

int N, C[MAX];
vector<int> G[MAX];
priority_queue<ll> A, B;

void dfs(int u, int p)
{
    C[u] = 1;
    for (int v : G[u]) {
        if (v == p) continue;
        dfs(v, u);
        C[u] += C[v];
        B.push((ll) C[v] * (N - C[v]));
    }
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    cin >> N;
    for (int i = 1; i < N; i++) {
        int u, v; cin >> u >> v;
        G[u].push_back(v);
        G[v].push_back(u);
    }
    for (int i = 1; i < N; i++) {
        int n; cin >> n;
        A.push(-n);
    }

    dfs(1, 0);

    ll ans = 0;
    for (int i = 0; i < N-1; i++) {
        ll a = A.top(), b = B.top(); A.pop(); B.pop();
        ans = (ans - a * b) % MOD;
    }
    cout << ans << '\n';

    return 0;
}
